### ISSUE TITLE
Group Polygons

### ISSUE THREAD

**Issue by chasemaier** [xyonix/vannot#22]
In almost all of our current use-case we need to conceptually group various polygons to related objects.

E.g. imagine we had labels "top" "bottom" "left" "right" for a type of object, but there could potentially be 4+ of those objects in the frame at the same time. When inspecting the output data we need some way to know which set of the 4 lables represent EACH overall object.

Initial though is that we (optionally) adjust the object structure from:
```
{
    "id": 4,
    "title": "top",
    "color": "#ecff00"
}
```

to

```
{
    "id": 4,
    "title": "top",
    "color": "#ecff00",
    "group": "4574e83c-606d-11e8-9c2d-fa7ae01bbebc"
}
```

The new group key would be arbitrary (e.g. hash/timestamp/incrementing integer/etc...) but would be common among polygons belonging to the same overall object so they could be grouped during post-processing of the output data.

The groupings would be unique WITHIN a frame.

We can potentially solve this with long "nested" label strings but it ends up being a multiplicatively increasing number of labels: `[number of overall object types] * [number of object segments] * [max number of objects that could appear in any given frame]` which would not be fun to scroll through in a dropdown. At least adding the grouping option would eliminate the last multiplication from that number.

If the UI knew about object hierarchies, that would be cool, but what I described above doesn't require that -- we'd just need to add some way to group existing polygons into arbitrary groups.

**Comment by clint-tseng** [comment]
I think I am following, but I'm not very certain so I want to confirm: are you essentially asking for the objects in the timeline below to be groupable into folders (and of course respective changes to the data itself)? Nothing to do with the actual individual drawn shapes, yes?

So instead of having eg "Arm/lower" "Arm/upper" "Leg/lower" "Leg/upper" as four top-level assignable objects below and in the dropdown, you'd have group Arm with upper and lower and same for Leg?

If so, is there an ask for multi-level nesting?

Even if I'm misunderstanding, I think once I grok what you're referring to it's a reasonably small ask/implementation time, I can probably turn it around quite quickly.

**Comment by chasemaier** [comment]
The hierarchy would be great, but is secondary to the grouping.

So, for example, with your examples, assuming we have the "Arm/lower" "Arm/upper" "Leg/lower" "Leg/upper" type labels:

- Imagine we have images of people, we need to have "Arm/lower" and "Arm/upper" polygons independently marked on those areas.
- However, people have multiple arms, and maybe there are many people in the image, so in the exported data we need to know that "this lower arm polygon is associated with this upper arm polygon and forms a whole inter-connected 'arm'", so those would be "grouped" together in some way.
- That way we'd be able to distinguish complete arms, even if there were multiple arms on a person or multiple people in the image.
- This could actually also apply with the same/single label too -- e.g. if there was a single top-level "torso" label, it may be necessary to "group" two separate "torso" polygons that are part of the same person if they are carrying/have some object in from of them that obscures a portion of their torso.
- FUTUE: I could see a need to have this nested (e.g. arm + leg + torso groups are further grouped into "person" group to identify individual people as a whole), but we do not currently have such a use-case.

If the UI was able to support the hierarchy of this type it would be optimal, but as long as the exported data includes the grouping data, that's the main thing that is needed (e.g. even if it was just a "click 3 polygons and click a "group" button" or something).

Hopefully that helps... a least a little. Let me know if I can further clarify. Thanks!

**Comment by alex9311** [comment]
maybe this helps:
Instead of defining that all forearms and hands are part of "arm", we need to say forearm x and hand y go together. 

In object oriented speak, we need to group instances of classes together, rather than group classes together

**Comment by clint-tseng** [comment]
It sounds like there are maybe multiple things going on here:

* The distinction between the class of an object (arm) and one instance of that object (this particular left arm).
* The grouping of multiple instances (of possibly various classes) into an unrelated class (this arm and that arm and this torso is this single person).
* This all is assuming that the distinction between eg upper and lower arms, as opposed to simply understanding that these two polygons together form a whole arm, is something you are interested in preserving.

I see what you mean that with a sufficiently complex naming scheme this would be doable but extremely painful. But what I'm sensing is that even that isn't necessarily quite correct—since the second bullet implies a not-purely-hierarchical semantic relationship (this arm belongs to this person, yes, but it's also still an arm and that's still key information).

We do, of course, already have the ability to attach multiple polygons per frame to any particular object. But if we are interested in tracking which polygons relate to which instance of that object then this becomes insufficient, unless we form a separate object for each instance, at which point we have too many objects to be manageable.

I'm trying to think of a way to organize these concepts in the interface in a way that would actually make sense to the annotating user. In the meantime, my questions are:

* I sense that you have some expedient short-term solution proposed but I'm not sure I understand how to translate it into the UI.
* Are you interested in the constancy of instances between frames? Or just in separating/grouping instances within each frame?

**Comment by chasemaier** [comment]
`But if we are interested in tracking which polygons relate to which instance of that object then this becomes insufficient` <== yeah, that's the problem that we're attempting to solve with this

To your last two questions:
- Agreed. The UI part is the aspect I wasn't sure about either.
  - The most simply (though not nice UX) would be to just allow the user to click/select multiple polygons and click a "group these" button that would just inject a new unique groupId value into those polygon objects. Keeping track of that state / allowing undo / surfacing that they are grouped in the UI seems sightly difficult.
  - Maybe just having some widget/box off on the side that showed the groups on the current frame and let you see which objects were part of which group / break them apart / etc... would be sufficient. That seems like the best UX I can come up with on the spot.
- We only care WITHIN a specific frame at this time. The same object across different frames do not need associated with each other in any way.

**Comment by ctusk** [comment]
You can achieve all of this by simply allowing nested labels

person_1
  - arm
        -upper arm
       - lower arm
person_2
  - arm
        -upper arm
       - lower arm

I would not make it more complex than absolutely necessary to achieve the desired semantic grouping via a combination of naming scheme and hierarchy. The most important issue is the user experience. While labeling you do not want a hierarchy that is nested too deep to comfortably work with and at the same time you don't want too many top level labels so the list you choose from does not get too long.

**Comment by clint-tseng** [comment]
Okay; not needing constancy between frames helps a lot a lot.

There is an alternative approach here in the short term, I think, which would be the ability to draw occlusion shapes. So each polygon would still be considered an instance, but it would be possible to chop it into multiple segments by the overlay of one or more occluding polygons. I'd long considered such a concept, but for slightly different reasons, and I'm not sure it's any more natural a solution here than direct grouping of some kind, either at the polygon or the label level.

I will give this some more thought, but I can't promise an immediate turnaround on it.

**Comment by ctusk** [comment]
It's not really about occlusion though, it is literally about objects having labeled sub-objects having labeled sub-objects, although a single level hierarchy would already help. If each label had a parent label the problem would be solved.

**Comment by clint-tseng** [comment]
As I am understanding it: the issue with using a purely label-based solution, though, is a conflation between class <=> class hierarchies and class <=> instance relationships; if, for instance, there are three torsos in the frame, and I create three sublabels "torso A"…"C" under "torso," then I advance to the next keyframe and there are only two torsos, one of which wasn't present in the previous keyframe, it becomes unclear whether I'm supposed to create "torso D" or not. It becomes even more confusing if "torso" is actually already a sublabel within "person".

Of course, these issues can be solved with case-specific instructions, but I just want to be sure we are all solving the same problem here. I am, as always, happy to do (or not do) whatever.

**Comment by alex9311** [comment]
we're still complicating the issue, we just need a way to group objects inside a single frame. The grouping does not need to be meaningful. Assuming we have a picture with two arms, where the types of objects we can draw are hands and forearms.

current:
```[
{
    "id": 1,
    "title": "hand",
    "color": "#ecff00"
},
{
    "id": 2,
    "title": "hand",
    "color": "#ecff00"
},
{
    "id": 3,
    "title": "forearm",
    "color": "#FFFFFF"
},
{
    "id": 4,
    "title": "forearm",
    "color": "#FFFFFF"
}
```
desired:
```[
{
    "id": 1,
    "title": "hand",
    "color": "#ecff00",
    "group": "4574e83c-606d-11e8-9c2d-fa7ae01bbebc"
},
{
    "id": 2,
    "title": "hand",
    "color": "#ecff00",
    "group": "xxxxxxxxx"

},
{
    "id": 3,
    "title": "forearm",
    "color": "#FFFFFF",
    "group": "4574e83c-606d-11e8-9c2d-fa7ae01bbebc"
},
{
    "id": 4,
    "title": "forearm",
    "color": "#FFFFFF",
    "group": "xxxxxxxxx"
}
```

the `group` field can be randomly generated. The group does not need to persist across frames

**Comment by ctusk** [comment]
You would still have the UI issue of having too many labels to deal with a the top level if you don't implement the hierarchy. And we do want to be able to carry instances over between frames in the tool. So Person1 in frame1 should be Person1 in frame 10 if it is indeed the same person. 

I would solve the grouping by enabling hierarchical labels and leave the distinction of instances to the semantics of the label text ( to be determined on an application case by application case basis ).

**Comment by clint-tseng** [comment]
That seems likely, but my realtime/business hours schedule is challenging for the next couple of weeks.

**Comment by clint-tseng** [comment]
**worklist**:

~~1. add partial shapes section to points selection mode
    1a. add expand selection button (which should highlight the remaining points)
2. add instance section to whole-shapes toolbar
    2a. add "form instance group" button to whole-shape selection mode~~
    2b. add visual indication of whether a polygon is part of an instance or not
    ~~2c. add "break instance group" button to whole-shape toolbar (which should select the broken polygons)~~
3. instance classes
    3a. configuration option to show optional freeform instance name field
    3b. configuration option to preset available instance labels
    3c. add autocomplete to freeform instance name field
4. workflow warnings
    4a. configuration option to display warning when a polygon is expected to be part of an instance but isn't
        (only when not selected?)
    4b. configuration option to display warning when any polygon is not part of an instance
5. hierarchical object grouping
    5a. add folder button makes folder
    5b. drag/drop manipulation

**Comment by clint-tseng** [comment]
n.b. i am using `instanceId` instead of `groupId` on the `shape` object.

**Comment by clint-tseng** [comment]
There is enough here now that people can probably try it out and give preliminary feedback; keep in mind of course that there is much of the worklist above to get through still and more UI things to implement that should smooth things over. But as utilitarian as it is now and as much improvement as can be made, it _does_ do everything you need I think.

**Comment by clint-tseng** [comment]
I have a plan for a better visual indication of the groupings that I'm wrapping up; it should be pushed sometime tomorrow.

**Comment by clint-tseng** [comment]
Okay, I have pushed a new version that implements a far nicer visual representation for instance groups. There's still a lot more to do on them, as discussed above and as you can see in the checklist, but I believe it does do everything you need for now, and I'm happy with how it's done for the moment. Please give it a try sometime and let me know what you think. I'm also happy to record another video to demonstrate it, but I can't do that until tomorrow.

**Comment by chasemaier** [comment]
:cool: @clint-tseng, I had a chance to test this on the branch today. UI / functionality seems sufficient for what we need to move forward. :shipit: 

If you feel like it's at a good stable point, maybe we can merge to master and continue section 3+ as a separate chunk of work.

**Comment by clint-tseng** [comment]
Great to hear. I'll take another look tomorrow and see if I'm happy to merge and prioritize that, with section 3+ as separate tickets as you suggested. Thanks!

**Comment by clint-tseng** [comment]
Commits have been merged and tickets #25, #26, #27, and #28 have been created to track the remaining open issues discussed here.
